### PR TITLE
Added donate functionality to the back-end and Front-end

### PR DESCRIPTION
**Pull request by SarKurd**

### CODE REVIEW

File: api/src/donation/donate.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
+const mollieClient = require("@mollie/api-client")({
+    apiKey: "test_cc4R8u25pJ7CH7QvRhqHPNtqqMBeWp" //test API key
```
Comment (isaachinman) [inline api/src/donation/donate.js]: Should definitely be an env var, even for test keys.

File: api/src/app.js [hunk 2]
```diff
@@ -72,4 +73,15 @@ app.get("/get-applicant", (req, res) => {
         );
 });
 
+app.post("/donate", (req, res) => {
+    donate(req.body, res);
+});
+
+app.get("/", (req, res) => {
```
Comment (isaachinman) [inline api/src/app.js]: Why is this GET route at the root? Also, users can get the status of an order with just the order ID, and nothing else (no authentication at all)?

File: api/src/app.js [hunk 3]
```diff
@@ -72,4 +73,15 @@ app.get("/get-applicant", (req, res) => {
         );
 });
 
+app.post("/donate", (req, res) => {
+    donate(req.body, res);
```
Comment (isaachinman) [inline api/src/app.js]: Would be good to know what this `req.body` object is supposed to look like.

File: api/src/donation/donate.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+});
+
+function donate({ method, amount, description }, res) {
+    const orderId = new Date().getTime();
+
+    mollieClient.payments
+        .create({
+            amount: { value: `${amount}.00`, currency: "EUR" },
```
Comment (isaachinman) [inline api/src/donation/donate.js]: Why not `toFixed(2)`?

File: api/src/donation/donate.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+            description: description,
+            redirectUrl: `http://localhost:3000/`,
```
Comment (isaachinman) [inline api/src/donation/donate.js]: This should be an env var as well. I assume it cannot be a relative URL in the case of iDEAL callbacks from other pages.

File: api/src/donation/donate.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+            webhookUrl: `https://sarbast.heroku.com`,
```
Comment (isaachinman) [inline api/src/donation/donate.js]: Should probably be an env var as well to support different environments (local, staging, prod).

File: api/src/app.js (hunk 2 above)
Comment (SarKurd) [reply api/src/app.js]: The GET route is to return the user to home but yes it's unnecessary
I'm sorry but what kind of authentication do mean? you mean express-session?

File: api/src/donation/donate.js (hunk 1 above)
Comment (SarKurd) [reply api/src/donation/donate.js]: Agreed, forgot to do that :-D

File: api/src/app.js (hunk 3 above)
Comment (SarKurd) [reply api/src/app.js]: You can check the `donate.js` but will update it

File: api/src/donation/donate.js (hunk 4 above)
Comment (SarKurd) [reply api/src/donation/donate.js]: Another thing i forgot, since i was playing around the code

File: api/src/donation/donate.js (hunk 5 above)
Comment (SarKurd) [reply api/src/donation/donate.js]: Will update that

File: api/src/app.js (hunk 2 above)
Comment (isaachinman) [reply api/src/app.js]: > return the user to home 

I don't understand what this means at all.

By authentication, I mean that usually we would restrict access to order/payment information by user. If we don't have user authentication, we'd use tokens in query parameters, etc.

Comment (SarKurd) [reply api/src/app.js]: Redirect user after the payment :-D

I will try to implement that

**Comment by isaachinman** [comment]
Perhaps @Fadi-Rasheed can check this out, as `email-crypto.js` was his work I believe.

**Comment by SarKurd** [comment]
I had to change the name so i can use it in donate.js too, now we can use it in other places too, i'll request him to review it

File: api/src/donation/donate.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    mollieClient.payments
+        .create({
+            amount: {
+                value: Number.parseFloat(amount).toFixed(2),
+                currency: "EUR"
+            },
+            description: description,
+            redirectUrl: redirectUrlFull,
+            webhookUrl: `https://sarbast.heroku.com`,
+            method: method,
```
Comment (M3kH) [inline api/src/donation/donate.js]: Not needed `method: method` => `method,` is just enough.

File: api/src/donation/donate.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+    mollieClient.payments
+        .create({
+            amount: {
+                value: Number.parseFloat(amount).toFixed(2),
+                currency: "EUR"
+            },
+            description: description,
+            redirectUrl: redirectUrlFull,
+            webhookUrl: `https://sarbast.heroku.com`,
```
Comment (M3kH) [inline api/src/donation/donate.js]: This needs to be an `env` variable.

File: api/src/donation/donate.js [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    const encryptedOrderId = encryptData(orderId.toString());
+    const redirectUrlFull = urljoin(baseURL, "?orderid=", encryptedOrderId);
+
+    mollieClient.payments
+        .create({
+            amount: {
+                value: Number.parseFloat(amount).toFixed(2),
+                currency: "EUR"
+            },
+            description: description,
```
Comment (M3kH) [inline api/src/donation/donate.js]: ` description: description` => `description`

File: api/src/donation/donate.js [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+            redirectUrl: redirectUrlFull,
```
Comment (M3kH) [inline api/src/donation/donate.js]: Rename `redirectUrlFull` to `redirectUrl` and also here `redirectUrlFull,` would be just enough.

File: api/src/donation/donate.js [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        })
+        .then(payment => {
+            res.redirect(payment.getPaymentUrl());
+        })
+        .catch(error => {
+            res.json({ error: error });
+        });
+}
+
+function paymentStatus(encryptedOrderId) {
```
Comment (M3kH) [inline api/src/donation/donate.js]: I can't see where this is use, do we miss still one endpoint in `middlewares`?

File: api/src/donation/donate.js [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    const orderId = decryptData(encryptedOrderId);
+    return new Promise((resolve, reject) => {
+        mollieClient.payments
+            .all()
+            .then(payments => {
+                payments.map(payment => {
+                    if (payment.metadata.orderId === parseInt(orderId)) {
+                        if (payment.isPaid()) {
+                            resolve("Thanks for the donation");
```
Comment (M3kH) [inline api/src/donation/donate.js]: This should return I guess an redirect to a page where we say show our grateful message :-)

File: api/src/donation/donate.js (hunk 11 above)
Comment (SarKurd) [reply api/src/donation/donate.js]: I haven't used it yet, i planned to make the `redirectURL` to send back the user to the homepage and the get request has the encrypted `orderId` as a query and show a message to the donator based on the payment status

```
app.get("/", (req, res) => {
  const orderId = decryptData(req.query.orderId);
  if (orderId) {
    paymentStatus(orderId)
      .then(msg => res.status(200).json({ msg }))
      .catch(err => res.status(500).json({ err }));
  }
});
```

let me know if this works

File: api/src/donation/donate.js (hunk 12 above)
Comment (SarKurd) [reply api/src/donation/donate.js]: I answered in above reply

**Comment by SarKurd** [comment]
Here what i did to show the donater a message base on the payment status

`redirectURL` will redirect the donater to the homepage and pass the encrypted `orderId`
Vue home page(`headerNavDefault` which is the component for non-mobile version broswer :/) will query the `orderId` and  fetch it to get the status of the donation

**code**

`donate.js`

```javascript
mollieClient.payments
    .create({
      amount: {
        value: Number.parseFloat(amount).toFixed(2),
        currency: "EUR"
      },
      description,
      redirectUrl,
      webhookUrl: process.env.WEBHOOK_URL,
      method,
      metadata: { orderId }
    })
    .then(payment => {
      res.redirect(payment.getPaymentUrl());
    })
    .catch(error => {
      res.json({ error: error });
    });
```
**`HeaderNavMobile.vue`**

```javascript
method: {
.
.
.
paymentStatus: function() {
      const baseURL = new URL(
        process.env.GET_DONATION_STATUS_URL ||
          "http://localhost:3005/donationstatus"
      );

      const param = { orderid: thos.$route.query.orderid };

      baseURL.search = new URLSearchParams(params);

      fetch(baseURL)
        .then(response => console.log(response))
        .catch(err => console.log(err));
    }
  },
  mounted() {
    this.paymentStatus();
  }

```

**`app.js`** the route that will return the payment status

```javascript
app.get("/donationstatus", (req, res) => {
  const orderId = decryptData(req.query.orderId);
  if (orderId) {
    paymentStatus(orderId)
      .then(msg => res.status(200).json({ msg }))
      .catch(err => res.status(500).json({ err }));
  }
});

```

please let me know if this is the right way to do it

**Comment by isaachinman** [comment]
Why do you want to take this approach in the first place? What is wrong with the user receiving donation status via email, or via payment confirmation from their bank? @mkruijt is there a formal specification for this payment functionality?

**Comment by SarKurd** [comment]
Because they want to show a message to the donator :-D

File: api/src/donation/donate.js [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+                currency: "EUR"
+            },
+            description,
+            redirectUrl,
+            webhookUrl: process.env.WEBHOOK_URL,
+            method,
+            metadata: { orderId }
+        })
+        .then(payment => {
+            res.redirect(payment.getPaymentUrl());
```
Comment (isaachinman) [inline api/src/donation/donate.js]: Change to:

    res.json({
        paymentURL: payment.getPaymentUrl()
    });

File: components/Donate.vue [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+                    "Content-Type": "application/json"
+                },
+                body: JSON.stringify({
+                    method: this.method,
+                    amount: this.amount,
+                    description: this.description
+                })
+            })
+                .then(res => res.json())
+                .then(json => console.log(json));
```
Comment (isaachinman) [inline components/Donate.vue]: Change to:

    .then(json => {
      window.location = json.paymentURL
    });

**Comment by isaachinman** [comment]
Please support the `decryptEmail` function so that other people's ongoing work is not broken. We will refactor this out in the future:

    module.exports = {
      decryptData: decryptData,
      decryptEmail: decryptData,
    }

**Comment by isaachinman** [comment]
Also, there are conflicts in `package-lock` that need to be resolved.

**Comment by isaachinman** [comment]
@SarKurd Let me know when this is ready for review/merge.

**Comment by SarKurd** [comment]
Issac it's ready to be reviewed then i resolve the conflicts

File: api/package.json [hunk 15]
```diff
@@ -26,10 +26,12 @@
     "webpack-cli": "^2.0.13"
   },
   "dependencies": {
-    "aws-sdk": "^2.300.0",
+    "@mollie/api-client": "^2.0.1",
+    "aws-sdk": "^2.291.0",
```
Comment (isaachinman) [inline api/package.json]: Careful of dependencies here. This is also the cause of your conflict within the `package.lock`, I believe.

Comment (SarKurd) [reply api/package.json]: Someone else added 2.300 so i removed and put the original one back, i guess it wasn't a good idea

**Pull request merged**